File: src/controllers/artwork.ts
import { Context } from "hono";
import { getPixivAccessToken, pixivGet } from "../clients/pixiv";
import { enrichArtworkResponseWithResolvedUrls } from "../lib/pixivImageResolver";

export async function artworkController(c: Context) {
  const idRaw = c.req.query("id");
  const illustId = Number(idRaw);

  if (!Number.isInteger(illustId) || illustId <= 0) {
    return c.json({ error: "Invalid artwork id" }, 400);
  }

  const refreshToken = process.env.PIXIV_REFRESH_TOKEN;
  if (!refreshToken) {
    return c.json({ error: "Missing env: PIXIV_REFRESH_TOKEN" }, 500);
  }

  try {
    const accessToken = await getPixivAccessToken(refreshToken);
    const response = await pixivGet(
      "/v1/illust/detail",
      { illust_id: String(illustId) },
      accessToken,
    );
    const baseUrl = new URL(c.req.url).origin;
    const enriched = enrichArtworkResponseWithResolvedUrls(response, baseUrl);
    return c.json(enriched as object);
  } catch (error) {
    return c.json(
      { error: "Failed to get artwork", detail: (error as Error).message },
      500,
    );
  }
}

File: controller/artwork.go
package controller

import (
	"strconv"

	"github.com/gofiber/fiber/v2"
	"pixivhono/client"
	"pixivhono/config"
	"pixivhono/lib"
)

func ArtworkController(c *fiber.Ctx) error {
	idRaw := c.Query("id")
	illustId, err := strconv.Atoi(idRaw)

	if err != nil || illustId <= 0 {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "Invalid artwork id",
		})
	}

	refreshToken := config.GlobalConfig.PixivRefreshToken
	if refreshToken == "" {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error": "Missing env: PIXIV_REFRESH_TOKEN",
		})
	}

	accessToken, err := client.GetPixivAccessToken(refreshToken)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error":  "Failed to get artwork",
			"detail": err.Error(),
		})
	}

	params := map[string]string{
		"illust_id": strconv.Itoa(illustId),
	}

	resp, err := client.PixivGet("/v1/illust/detail", params, accessToken)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error":  "Failed to get artwork",
			"detail": err.Error(),
		})
	}

	baseUrl := c.BaseURL()
	enriched := lib.EnrichArtworkResponseWithResolvedUrls(resp, baseUrl)

	return c.JSON(enriched)
}
